diff --git a/wazo_auth/database/queries/saml_pysaml2_cache.py b/wazo_auth/database/queries/saml_pysaml2_cache.py
index c4831f0b..8efedad1 100644
--- a/wazo_auth/database/queries/saml_pysaml2_cache.py
+++ b/wazo_auth/database/queries/saml_pysaml2_cache.py
@@ -109,7 +117,7 @@ class SAMLPysaml2CacheDAO(filters.FilterMixin, BaseDAO):
 
         cni = code(name_id)
 
-        if self.get(name_id, entity_id):
+        if self.get(name_id, entity_id, False):
             search_filter = self.new_search_filter(name_id=cni, entity_id=entity_id)
             data = {
                 'name_id': cni,
